_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,769 @@
+
+
+### GitOps: Operation By Pull Request
+
+GitOps, Figure 5, is a pattern for managing infrastructure. The core idea of GitOps is that source control (doesn’t have to be git) should be the source of truth for configuration files  describing your infrastructure. Controllers can then monitor source control and automatically update your infrastructure as your config changes. This means to make a change (or undo a change) you just open a pull request.
+
+
+<table>
+  <tr>
+   <td>
+
+<img src="/images/2020-08-01-data-science-meets-devops/fig5.gitops.png" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 5 To push a new model for Label Bot we create a PR updating the config map storing the id of the Auto ML model we want to use. When the PR is merged, <a href="https://cloud.google.com/anthos-config-management/docs">Anthos Config Management(ACM</a>) automatically rolls out those changes to our GKE cluster. As a result, subsequent predictions are made using the new model. (Image courtesy of <a href="https://www.weave.works/blog/automate-kubernetes-with-gitops">Weaveworks</a>)
+   </td>
+  </tr>
+</table>
+
+
+
+### Putting It Together: Reconciler + GitOps = CI/CD for ML
+
+With that background out of the way, let’s dive into how we built CI/CD for ML by combining the Reconciler and GitOps patterns.
+
+There were three problems we needed to solve
+
+
+
+1. How do we compute the diff between the desired and actual state of the world?
+2. How do we affect the changes needed to make the actual state match the desired state?
+3. How do we build a control loop to continuously run 1 & 2?
+
+
+#### Computing Diffs
+
+To compute the diffs we just write lambdas that do exactly what we want. So in this case we wrote two lambas
+
+
+
+1. The [first lambda](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/Label_Microservice/go/cmd/automl/pkg/server/server.go#L109) determines whether we need to retrain based on the age of the most recent model
+2. The [second lambda](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/Label_Microservice/go/cmd/automl/pkg/server/server.go#L49) determines whether the model needs to be updated by comparing the most recently trained model to the model listed in a config map checked into source control.
+
+We wrap these lambdas in a simple web server and deploy on Kubernetes. One reason we chose this approach is because we wanted to rely on Kubernetes’ [git-sync](https://github.com/kubernetes/git-sync) to mirror our repository to a pod volume. This makes our lambdas super simple because all the git management is taken care of by a side-car running [git-sync](https://github.com/kubernetes/git-sync).
+
+
+#### Actuation
+
+To apply the changes necessary, we use Tekton to glue together various CLIs that we use to perform the various steps.
```

Comment:
Tekton and not kubeflow pipelines? Or tekton via kubeflow pipelines?

Reply:
If not KFP, why not?

Reply:
Added a section on "Why Tekton"